### PR TITLE
Use QR decomposition for computing pseudoinverse in fantasy strategies

### PR DESCRIPTION
**Pull request by Balandat**
More numerically stable. Also adds jitter if the matrix is singular.

### CODE REVIEW

File: gpytorch/models/exact_prediction_strategies.py [hunk 1]
```diff
@@ -196,27 +199,29 @@ def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_
...
         new_root[..., m:, n:] = schur_root
 
         # Use pseudo-inverse of Z as new inv root
-        # TODO: Replace pseudo-inverse calculation with something more stable than normal equations once
-        # one of torch.svd, torch.qr, or torch.pinverse works in batch mode.
-        cap_mat = new_root.transpose(-2, -1).matmul(new_root)
-        if cap_mat.requires_grad or new_root.requires_grad:
+        # TODO: Consider replacing pseudo-inverse calculation with torch.svd / torch.pinverse once these
+        # methods work in batch mode.
+        Q, R = torch.qr(new_root)
```
Comment (Balandat) [inline gpytorch/models/exact_prediction_strategies.py]: This is where the interesting stuff is, sorry for mushing this together with the black reformats.

File: gpytorch/models/exact_prediction_strategies.py [hunk 2]
```diff
@@ -196,27 +199,29 @@ def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_
...
+        Rdiag = torch.diagonal(R, dim1=-2, dim2=-1)
+        # if R is almost singular, add jitter (Rdiag is a view, so this works)
+        zeroish = Rdiag.abs() < 1e-6
+        if torch.any(zeroish):
+            Rdiag[zeroish] = 1e-6
+        if R.requires_grad or Q.requires_grad:
```
Comment (jacobrgardner) [inline gpytorch/models/exact_prediction_strategies.py]: `R` is already triangular, so we should be able to compute `R^{-1}Q^{\T}` with `torch.trtrs`.

Comment (jacobrgardner) [reply gpytorch/models/exact_prediction_strategies.py]: Sorry, `torch.traingular_solve` as it is now called :-)

File: gpytorch/models/exact_prediction_strategies.py [hunk 3]
```diff
@@ -196,27 +199,29 @@ def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_
...
-        new_root[..., :m, n:] = upper_right
         new_root[..., m:, :n] = lower_left
         new_root[..., m:, n:] = schur_root
 
         # Use pseudo-inverse of Z as new inv root
-        # TODO: Replace pseudo-inverse calculation with something more stable than normal equations once
-        # one of torch.svd, torch.qr, or torch.pinverse works in batch mode.
-        cap_mat = new_root.transpose(-2, -1).matmul(new_root)
-        if cap_mat.requires_grad or new_root.requires_grad:
+        # TODO: Consider replacing pseudo-inverse calculation with torch.svd / torch.pinverse once these
```
Comment (jacobrgardner) [inline gpytorch/models/exact_prediction_strategies.py]: regarding the todo comment, `torch.qr` should always be preferable to `torch.svd`, being much (probably ~10x) faster and comparably stable. `torch.pinverse` currently uses `torch.svd` internally, but this is a bad thing and it should use `qr`.

File: gpytorch/models/exact_prediction_strategies.py (hunk 2 above)
Comment (Balandat) [reply gpytorch/models/exact_prediction_strategies.py]: ha I was looking for that but didn't find it, and I didn't want to manually massage R into an LU factorization... thanks

**Comment by Balandat** [comment]
Changed this to fall back to old implementation when running on stable (not backporting the jitter though).

**Pull request merged**